### ISSUE TITLE
users are forced to upgrade their pip for 0.17

### ISSUE THREAD

**Issue by emmanuelle** [scikit-image/scikit-image#4674]
See for example #4667 and #4673, there seems to be a problem with `tifffile` in some `pip` versions (including versions as recent as 18.1, which is the version installed when I'm creating a venv - even if pip in my main environment is the most recent release). For a major and stable package like scikit-image, it is ungracious to ask users to upgrade their pip (some people might fear breaking other stuff). Should we temporarily revert https://github.com/scikit-image/scikit-image/pull/4235 to alleviate this issue?

**Comment by hmaarrfk** [comment]
I don't think it is unreasonable to ask users to update their install of pip.

Our instructions on the INSTALL.rst should point them to that. I think we should work on defining the scope and usecases we want to cover there.

In my opinion: ubuntu's decision to freeze pip ubuntu's choice. For scikit-image, they offer `python3-skimage`. If user's want bleeding edge packages, they should use up-to-date package managers.

Do we need tifffile? can we make it an optional dependency?

**Comment by emmanuelle** [comment]
Thanks for your perspective @hmaarrfk. I have to say I was a bit shocked when I had to upgrade my pip recently to install napari since it was the first time ever I had to upgrade pip to install something. I think there is a correlation in some people's mind between the fact to force upgrading some packages to recent versions and a poor attention to stability (usually I hate it when packages force me to upgrade numpy or jupyter). Most people have only one environment when they work.

**Comment by jni** [comment]
> since it was the first time ever I had to upgrade pip to install something.

@emmanuelle in general I agree with you, but I think the pip/PyPA team have been introducing changes very quickly and prodding users to upgrade. So I don't think this is all "on us". (Note that I generally think this is a good thing because Python packaging is a mess. But there will be growing pains as we are now seeing.)

I have a slight preference towards making tifffile optional than re-bundling it, since there was quite a bit of user confusion about why skimage tifffile behaved differently from tifffile. Having said that, I would then expect many users to wonder why `skimage.io` no longer reads tiffs properly. So we are in a hard situation.

I wonder whether @cgohlke might be able to comment on whether releasing `manylinux1` wheels for imagecodecs would be feasible, since that would solve the situation. Clearly it is not trivial since automatic compilation is failing...

Actually I think the tifffile documentation on PyPI is out of date, because it lists imagecodecs as an optional dependency, whereas the logs from #4667 suggest that it is being pulled in by tifffile. I think this is a recent thing so users might be able to install an earlier tifffile version to avoid pulling in a recent tifffile and thus imagecodecs.

Anyway, long story short, my preference is to continue to depend on tifffile and try to find a way to make pip do what we want. I'll point out that users' machines are capable of installing the manylinux2010 wheels, it's just that old pip doesn't know how to find them. See [here](https://pip.pypa.io/en/stable/news/) for the pip changelog and [PEP-571](https://www.python.org/dev/peps/pep-0571/) for the manylinux2010 standard. So it's a shame to go back to diverging tifffile versions just because users can't get the wheel they need.

**Comment by rfezzani** [comment]
I already given my opinion on this, and sorry to bother you with it, but I really feel that `skimage` can gain stability by focusing on image processing! So many issues and bugs are related to what I call "out of scope modules".
It would be, I think, a great improvement if we can find a way to deprecate `io`, `viewer` and `data` modules in favor of respectively `imageio`, `napari` and `skimage-data` for example.
This automatically reduces the number of dependencies, making the package lighter and better scoped.

**Comment by jni** [comment]
Oh, actually I'm totally in favour of this, @rfezzani! But we can't deprecate io in a patch release. =) Shall we deprecate it altogether in 0.18? (ie warn that it will go away in 0.20/1.0?) But that doesn't solve @emmanuelle's issue of wanting people to be able to install 0.17 on older pips.

**Comment by emmanuelle** [comment]
Not at all @rfezzani thanks for sharing your opinion, saying things twice or more is usually a good idea as we're all too busy and some information might slip though! I have mixed feelings about this:
- for the viewer I'm all for deprecating and directing users to napari / matplotlib / plotly. Anyway it's not something which is used a lot.
- for the data I might be convinced since it's mostly for the docs and for us developers (I find myself loading these images all the time but for us it wouldn't make a difference to import them from somewhere else). Anyway we would be the ones to maintain a skimage-data module.
- for I/O... meh. As you know `io.imread` is scikit-image's most popular function so I feel we would really let users down by removing this feature. But it's worth discussing it if everything could be done with `imageio`, or if we could simplify drastically our I/O logic to have thin wrappers around imageio. Do we have an open issue about this?

**Comment by rfezzani** [comment]
Of course @jni =), my message is off topic concerning this issue, sorry. I am more on the side of @hmaarrfk here: users needing up to date packages usually don't care to update their package managers =p

Reading this issue, and the other issues/PR concerning `pooch` just made me write this comment for "out of scope" modules deprecation.

@emmanuelle, I am not sure, but isn't io.imread` calling by default the `imageio` plugin? Deprecating the `io` module can come with a documentation section describing the best practice for this topic (calling `imageio`, `tifffile` and others for more exotic formats...)

**Comment by sciunto** [comment]
>     * for I/O... meh. As you know `io.imread` is scikit-image's most popular function so I feel we would really let users down by removing this feature. But it's worth discussing it if everything could be done with `imageio`, or if we could simplify drastically our I/O logic to have thin wrappers around imageio. Do we have an open issue about this?

#1605

**Comment by hmaarrfk** [comment]
I think for the last year, pypa has been adding warnings to users to upgrade their pip. Unfortunately, Ubuntu's 18.04 pip is 2 years old ;).

That said, Ubuntu 20.04 provides a much more modern pip:
```
pip 20.0.2 from /usr/lib/python3/dist-packages/pip (python 3.8)
```
So I think this issue will come to pass.

Depending on modules that will only ship manylinux2010 is an inevitability.

While not every dev seems every problem, I think collectively, we've seen enough challenges to write a good user install guide in light of modern packaging practices. We *need* to update our install instructions to reflect these. People's input on https://github.com/scikit-image/scikit-image/issues/4672 would be greatly appreciated on that.

**Comment by Leon-Sam** [comment]
I'll chime in as well, since I created isssue #4673 . As a novice python user, I'm not too concerned about the exact pip version that is on my system/project. 

My larger concern is just making sure that the project the I am working runs and can be easily reproduced by others. If the instructions for scikit-image said to make sure the pip version is greater than X.X. Then I'll happily upgrade the version.

Cheers!

**Comment by hmaarrfk** [comment]
^^^ Thanks for your input. Super useful actually.

**Comment by emmanuelle** [comment]
Yes thanks for your input @Leon-Sam. Given the fact that the pip folks have encouraged upgrading and it's not only us, I'm going to close this issue, the follow up will be in #4672 (improving installation instructions).
